README.md
python-dynamic-table
====================

Python module for creating text table respresentations in various table. Including Text, CSV, and HTML

============
Introduction
============
This module provides a way to build/create, and print tables.  So you can 
Create a text table, add rows, and then print them in a human readable text
format.  It also supports html and csv renderers.  So if you build a text
table you can print it in html or csv format.

[NOTE] This was written quickly, but could use some additional features such
as:
 * Ability to change text justifcation for text and html tables by column
 * Ability to change the background cell color for html tables
 * Ability to print an html table in html5 supported ways with inline css etc...
 * In other words... this is heavily a work in progress

The interface to dynamic table is pretty straight forward and all of the parameters
passed to the general functions, such as:
 * add_row
 * set_table_type
 * set_col_widths
 * set_col_names
 * set_table_renderer

If you want to write the table to a file instead of stdout (the default) you have
to do so when instantiating the Table class.  Like so:
```
 my_file=open('/path/to/file','rw')
 my_table=Table(output=my_file)
```

Doing this will ensure that any print functions get written to the file instead of
stdout.

You can also modify this behavior by defining output as 'String' and then 
the print functions will build the table in the variable "built_buffer", this does
not get cleared automatically. When done with the data, call the function empty_output().
[NOTE] This way of doing this is deprecated in favor of calling str(Table). However still
useful if building adhoc tables I guess.
For example with adhoc table:
```
 my_table=Table(output='String')
 my_table.set_col_names(['Col1','Col2'])
 my_table.print_header()
 my_table.print_row(['a','b'])
 my_table.print_row(['c','d'])
 table_text=my_table.built_buffer
 my_table.empty_output()
```

Example of rendering the table to a string:
```
 my_table=Table()
 my_table.set_col_names(['Col1','Col2'])
 my_table.add_row(['a','b'])
 my_table.add_row(['c','d'])
 my_string=str(my_table)
```

=========
Renderers
=========

All Table objects have to have a Render type object. This can be changed at any time
after the creation of the object using the Table function set_table_renderer. The
default renderer if nothing is passed is RenderText. 

All Render classes must have the folloing functions (as they are called by Table)
the parameters for each differs for each type of table (all return string types):
  * print_header - Prints just the header row/column names
  * print_row    - Prints a single adhoc row
  * print_rows   - Prints all the rows in the table
  * print_table  - Prints the entire table

There are extra functions for the following renderers:
 'RenderText'
  * print_footer - Prints the footer of a text table

Supported parameters for print_table() for each renderer and defaults are below:
(Passed to the Renderer's __init__ functoin):
RenderText
----------
 ```
  * indent=0
  * borderless=False
  * color_disabled=False
  * padding=def_padding                    - def_padding=0
  * padding_char=def_txt_padding_char      - def_padding_char=' '
  * fill_char=def_txt_fill_char            - def_fill_char=' '
  * h_border_char=def_txt_horz_border_char - def_horz_border_char='-'
  * v_border_char=def_txt_vert_border_char - def_ver_border_char='|'
  * v_border_char=def_txt_vert_border_char - def_ver_border_char='|'
  * col_sep_char=def_txt_sep_char          - def_sep_char='|'
```
RenderCSV
---------
```
  * sep_char=def_sep_char          - def_sep_char=','
```

RenderHTML
----------
  [NOTE] this Renderer make use of Table's row_render_opts dict. Which means
         when calling Table's add_row method you can pass in a dict with the
         following keys/values:
           html_row_attr:  String HTMl row attribute/s to add to the 'tr' tag
           html_cell_attr: List of HTML cell attribute/s to add to 'td'/'th'
                           tags
```                           
  * color_disabled=False
  * table_attr=''
  * thead_attr=''
  * tbody_attr=''
```

Quick examples
--------------
```python
  from dynamic_table import *
  #Render the table indented 5 spaces, and cells padded 3 chars
  my_table=Table(RenderText(indent=5,padding=3))
  my_table.set_col_names(['Col1','Col2'])
  my_table.add_row(['a','b'])
  my_table.add_row(['c','d'])
  my_table.render()
```
```python
  #Render the same table above but in html with with a css class of "table":
  my_table.set_table_renderer(RenderHTML(table_attr='class="table"'))
  my_table.render()
```
```python
  #Render the same table, but now in csv pipe '|' separated
  my_table.set_table_renderer(RenderCSV(sep_char='|'))
  my_table.render()
  #Assign the built table to a variable:
  built_table=str(my_table)
```

=============
Table Filters
=============
Table objects can also be filtered using a set rules. This is done through
the use of TableFilter objects. They don't have any required parameters, but
can take an optional filter_txt string when being created.

Filter Expression Syntax (items in parenthesis are optional):
```
  ([column ids]);[column id][operator][comparison value];\ 
                 ([column id][operator][comparison value])...
```

Values for above:
```
  [column ids]:       comma separated list of columns ids. Ranges are
                      specified with a "-". e.g. 1-3. Ending the column ids
                      with a '-' will be from that column id on the left of
                      '-' to the end. Such as: "4-" would be from column 4 to
                      the end.
  [column id]:        Single column id
  [operator]:         One of: '>','<','>=','<=','!=','=','!/','/' which should
                      be self explanatory. '/' is a "contains" operator.
  [comparison value]: Value to compare against the column in [column id]. This
                      can be a string, number, or even a date.
```

If the first rule is a comma separated list it is referred to as a column
rule, all others are considered row rules. Additional row rules all have to be
true for a row to pass the filter.

Filter expression examples:
```
  #Only print columns 1,3 and 4 and only rows where column 1 is after
  #"2014-07-30 12:00:00"
  1,3,4;1>"2014-07-30 12:00:00"
  #Only print rows where column 4 is "Active" and column 2 has "tds" in it
  4=Active;2/tds
```

If not providing a filter_txt when creating a TableFilter object, you can add
more rules using the add_row_rule function and the set_col_rule. Which can be
useful if the [comparison value] contains a semi-colon in it.

TableFilter provides the following useful functions:
```
  filter_table(table):    returns a Table object that has been filtered
                          according to the rules in the TableFilter
  set_col_rule(col_rule): Add a column rule. (String, comma separated ids) See
                          Filter Expression Syntax, above
  add_row_rule(row_rule): Add a row rule. (String). See Filter Expression
                          Syntax above.
```                          

Examples of filtering:
```python
  #Create a table with some data in it:
  thing=dynamic_table.Table()
  thing.set_col_names(['Col1','Col2','Col3','Miscellaneous'])
  thing.add_row(['abc','blah','boo','stuff'])
  thing.add_row(['def','blah','boo','stuff'])
  thing.add_row(['ghi','blah','boo',''])
  thing.add_row(['ghi','blah','boo',''])
  thing.add_row(['jkl','blah','boo'])
  thing.add_row(['20140801 4:00:00','Im colored','right?'],['','red'])
  thing.add_row(['2013-04-21 1:00:00','','Yay Multicolors'],['','','bg_brown,black'])
  thing.add_row(['2014-07-30 12:00:00','a'])
  thing.add_row(['2014-07-30 16:00:00','b','c'],['green'])
  #Create a Table Filter to only have column 1 and 3 to the end and where
  # column 1 is after the date/time '2014-07-30 12:00:00'
  tf=dynamic_table.TableFilter(filter_txt='1,3-;1>2014-07-30 12:00:00')
  #Filter the table and render it
  tf.filter_table(thing).render()
```
```python
  #Create a filter that only show column 1,3 and 4, and where column 1 is
  # before '2014-07-30 12:00:00'
  tf=dynamic_table.TableFilter(filter_txt='1,3,4;1<2014-07-30 12:00:00')
  #Create a Table with the filter above that filters as it is being added
  thing=dynamic_table.Table(table_filter=tf)
  #Add some data and column names:
  thing.set_col_names(['Col1','Col2','Col3','Miscellaneous'])
  thing.add_row(['abc','blah','boo','stuff'])
  thing.add_row(['def','blah','boo','stuff'])
  thing.add_row(['ghi','blah','boo',''])
  thing.add_row(['jkl','blah','boo'])
  thing.add_row(['20140801 4:00:00','Im colored','right?'],['','red'])
  thing.add_row(['2013-04-21 1:00:00','','Yay Multicolors'],['','','bg_brown,black'])
  thing.add_row(['2014-07-30 12:00:00','a'])
  thing.add_row(['2014-07-30 16:00:00','b','c'],['green'])
  thing.render()
```

To see more examples of how all this works out see the file: dynamic_table_examples.py


debian/changelog
python-dynamic-table (0.8.7-%DIST%1) %DIST%; urgency=low

  * Remove useless border,padding and bordless attributes from RenderHTML in
    favor of just using table_attr

 -- Dan Farnsworth <absltkaos@gmail.com>  Wed, 18 Feb 2015 00:47:00 -0700

python-dynamic-table (0.8.6-%DIST%1) %DIST%; urgency=low

  * Fixed minor bug where a non-string in a row can break rendering

 -- Dan Farnsworth <absltkaos@gmail.com>  Fri, 30 Jan 2015 07:00:00 -0700

python-dynamic-table (0.8.5-%DIST%1) %DIST%; urgency=low

  * Feature: Support a not contains operator for TableFilters '!/'

 -- Dan Farnsworth <absltkaos@gmail.com>  Wed, 19 Nov 2014 18:00:00 -0700

python-dynamic-table (0.8.4-%DIST%1) %DIST%; urgency=low

  * Bug Fix: RenderHTML threw error when rendered without header set

 -- Dan Farnsworth <absltkaos@gmail.com>  Wed,  5 Nov 2014 11:00:00 -0700

python-dynamic-table (0.8.3-%DIST%1) %DIST%; urgency=low

  * Bug Fix: Improper TableFilters when a single column is to display
  * Bug Fix: Force TableFilter row rules that are string comparisons so the
    cells are converted to str

 -- Dan Farnsworth <absltkaos@gmail.com>  Tue,  4 Nov 2014 10:42:00 -0700

python-dynamic-table (0.8.2-%DIST%1) %DIST%; urgency=low

  * Bug Fix: Improper handling when no column rule is specified

 -- Dan Farnsworth <absltkaos@gmail.com>  Thu, 18 Sep 2014 17:01:00 -0600

python-dynamic-table (0.8.1-%DIST%2) %DIST%; urgency=low

  * Added an additional option to Table.add_row to add arbitrary rendering
    options for Render objects to use.
  * Added features to RenderHTML to make use of the new row_render_opts
    created from add_row, so attributes can be added to <tr> and <td>/<th>
  * Added table_attr option to RenderHTML for setting <table> attributes

 -- Dan Farnsworth <absltkaos@gmail.com>  Wed, 27 Aug 2014 17:42:00 -0600

python-dynamic-table (0.8.0-%DIST%2) %DIST%; urgency=low

  * Applied minor bug patch around TableFilter class, not handling number
    comparisons correctly due to poor operator logic

 -- Dan Farnsworth <absltkaos@gmail.com>  Mon, 18 Aug 2014 16:19:00 -0600

python-dynamic-table (0.8.0-%DIST%1) %DIST%; urgency=low

  * Add in TableFilters
  * Fixed Miscellaneous Bugs
  * Refactored the Table class to split the rendering of the table from the
    data
  * Added in __len__ and __str__ magic functions
  * Documented new features, and all Classes and functions.

 -- Dan Farnsworth <absltkaos@gmail.com>  Fri,  1 Aug 2014 16:00:00 -0600

python-dynamic-table (0.3.1-%DIST%1) %DIST%; urgency=low

  * Add ability to store the table in memory 

 -- Dan Farnsworth <absltkaos@gmail.com>  Mon, 23 Sep 2013 15:51:00 -0600

python-dynamic-table (0.3.0-%DIST%1) %DIST%; urgency=low

  * Initial release.

 -- Dan Farnsworth <absltkaos@gmail.com>  Mon, 23 Apr 2013 13:45:00 -0600


dynamic_table_examples.py
#!/usr/bin/python
import dynamic_table

#Example with printing multiple variable length rows that have been added
#Then converting to an html table
print("==Example of a text table printing variable length rows.Then Printing in HTML==")
thing=dynamic_table.Table()
#Set the column headers
print("Setting column names: ['Col1','Col2','Col3','Miscellaneous']")
thing.set_col_names(['Col1','Col2','Col3','Miscellaneous'])
#Add a bunch of rows
print("Adding rows to table. some with too few columns, and some with more, as well as some colors")
thing.add_row(['a','blah','boo','stuff'])
thing.add_row(['a','blah','boo','stuff'])
thing.add_row(['a','blah','boo',''])
#Add a row that doesn't have enough columns (should auto fill and be empty
thing.add_row(['a','blah','boo'])
#Add some rows with color, the second column will have red text, other row will have two attributes set
thing.add_row(['c','Im colored','right?'],['','red'])
thing.add_row(['d','','Yay Multicolors'],['','','bg_brown,black'])
thing.render()
#Adding a row with an extra column should add another column with other cells being empty
thing.add_row(['b','hrmmmmmm','ahhh','stuffing','big empty'])
print("Printing Table")
thing.render()
print("Changing table to type: html")
thing.set_table_renderer(dynamic_table.RenderHTML())
print("Printing Table")
thing.render()
del(thing)

# header and table resizes based on set column lengths and padding etc...
print("\n==Example of a text table with set column widths, and modifying padding==")
thing=dynamic_table.Table(dynamic_table.RenderText(padding=2,indent=2))
print("Setting column names: ['Col1','Col2','Col3','Miscellaneous']")
thing.set_col_names(['Col1','Col2','Col3','Miscellaneous'])
print("Adding a row")
thing.add_row(['a','blah','boo','stuff'])
print("Printing table as is with dynamic columns and indentation")
thing.render()
print("Setting static column widths, with an extra column of width 6")
thing.set_col_widths([5,10,3,5,6])
print("Printing table with indentation, static columns, and padding")
thing.render()
del(thing)


#Multi table test
print("\n==Example of creating two tables to make sure that passed by reference values aren't shared==")
print("Making table: thing")
thing=dynamic_table.Table()
thing.set_col_names(['Col1','Col2','Col3','Miscellaneous'])
thing.add_row(['a','blah','boo','stuff'])
thing.add_row(['a','blah','boo','stuff'])
thing.add_row(['a','blah','boo',''])
thing.add_row(['a','blah','boo'])
thing.add_row(['b','hrmmmmmm','ahhh','stuffing','big empty'])
print("Printing table: thing")
thing.render()
print("Making table: my_table")
my_table=dynamic_table.Table()
my_table.add_row(['1','value1','value2','value3'])
my_table.add_row(['2','value1','value2','value3'])
my_table.add_row(['3','value1','value2','value3'])
my_table.add_row(['4','value1','value2','value3'])
print("Printing table: my_table")
my_table.render()
del(my_table)
del(thing)

#An Adhoc text table example
print("\n==Example of an adhoc printed text table with static columns==")
thing=dynamic_table.Table()
print("Setting column names: ['Col1','Col2','Col3','Miscellaneous']")
thing.set_col_names(['Col1','Col2','Col3','Miscellaneous'])
print("Setting static column widths")
thing.set_col_widths(['5','8','8','8'])
print("Printing header,rows and footer")
thing.print_header()
thing.print_row(['1','value1','value2','value3'])
thing.print_row(['2','value1','value2','value3'],['','green'])
thing.print_row(['3','value1','value2','value3'])
thing.print_row(['4','value1','value2','value3'])
thing.print_footer()
del(thing)

#Creating a csv table and printing in csv, csv with pipes, and text
print("\n==Example of creating a csv table, printing, printing with difference delimeter, and then printing in text==")
render_plain=dynamic_table.RenderCSV()
render_pipes=dynamic_table.RenderCSV(sep_char='|')
thing=dynamic_table.Table(render_plain)
#Set the column headers
print("Setting column names: ['Col1','Col2','Col3','Miscellaneous']")
thing.set_col_names(['Col1','Col2','Col3','Miscellaneous'])
#Add a bunch of rows
print("Adding some various rows")
thing.add_row(['a','blah','boo','stuff'])
thing.add_row(['a','blah','boo','stuff'])
thing.add_row(['a','blah','boo',''])
#Add a row that doesn't have enough columns (should auto fill and be empty
thing.add_row(['a','blah','boo'])
#Add some rows with color, the second column will have red text, other row will have two attributes set
thing.add_row(['c','Im colored','right?'],['','red'])
thing.add_row(['d','','Yay Multicolors'],['','','bg_brown,black'])
#Print standard CSV
print("Printing table as is")
thing.render()
print("Printing table with pipe chars")
thing.set_table_renderer(render_pipes)
thing.render()
print("Printing table in text")
thing.set_table_renderer(dynamic_table.RenderText())
thing.render()
print("Printing table in text with a bunch of custom modifiers like custom padding, padding chars, fill chars etc..")
thing.set_table_renderer(dynamic_table.RenderText(padding=1,indent=2,padding_char='_',fill_char='<'))
thing.render()

print("\n==Example of Creating a table, Creating a Filter, then applying the filter to the table==")
#Creating a table and filtering it
thing=dynamic_table.Table()
#Set the column headers
thing.set_col_names(['Col1','Col2','Col3','Miscellaneous'])
#Add a bunch of rows
thing.add_row(['abc','blah','boo','stuff'])
thing.add_row(['def','blah','boo','stuff'])
thing.add_row(['ghi','blah','boo',''])
thing.add_row(['jkl','blah','boo'])
thing.add_row(['20140801 4:00:00','Im colored','right?'],['','red'])
thing.add_row(['2013-04-21 1:00:00','','Yay Multicolors'],['','','bg_brown,black'])
thing.add_row(['2014-07-30 12:00:00','a'])
thing.add_row(['2014-07-30 16:00:00','b','c'],['green'])
print("Table before filtering")
thing.render()
#Create the Table filter
print("Creating a table filter that will only return columns 1, and 3 to the end. but only rows where column 1 is after: 2014-07-30 12:00:00")
tf=dynamic_table.TableFilter(filter_txt='1,3-;1>2014-07-30 12:00:00')
#Filter and render the table:
tf.filter_table(thing).render()
del(thing)
del(tf)

print("\n==Example of creating a filter and creating a table with the filter, so data is filtered as it is added==")
#Create a filter that only show column 1,3 and 4, and where column 1 is
# before '2014-07-30 12:00:00'
print("Creating a table filter that will only return columns 1, 3, and 4. but only rows where column 1 is before: 2014-07-30 12:00:00")
tf=dynamic_table.TableFilter(filter_txt='1,3,4;1<2014-07-30 12:00:00')
#Create a Table with the filter above that filters as it is being added
print("Creating table, and adding filter to it")
thing=dynamic_table.Table(table_filter=tf)
#Add some data and column names:
print("Adding some data to the table")
thing.set_col_names(['Col1','Col2','Col3','Miscellaneous'])
thing.add_row(['abc','blah','boo','stuff'])
thing.add_row(['def','blah','boo','stuff'])
thing.add_row(['ghi','blah','boo',''])
thing.add_row(['jkl','blah','boo'])
thing.add_row(['20140801 4:00:00','Im colored','right?'],['','red'])
thing.add_row(['2013-04-21 1:00:00','','Yay Multicolors'],['','','bg_brown,black'])
thing.add_row(['2014-07-30 12:00:00','a'])
thing.add_row(['2014-07-30 16:00:00','b','c'],['green'])
#Show the filtered table
print("Rendering the filtered data")
thing.render()
del(thing)

print("\n==Example of creating an HTML table with special attributes for table, tr and td tags==")
print("Creating HTML Renderer that will have a class attribute of 'tablesort'")
rend=dynamic_table.RenderHTML(table_attr='class="tablesort"')
thing=dynamic_table.Table(rend)
thing.set_col_names(['Col1','Col2','Col3','Miscellaneous'])
thing.add_row(['a','blah','boo','stuff'])
print("Create second row with a class attribute of 'my_class', and third cell with class attribute set to o_class")
thing.add_row(['a','blah','boo','stuff'],[],{'html_row_attr': 'class="my_class"', 'html_cell_attr': ['','','class="o_class"']})
thing.add_row(['a','blah','boo',''])
thing.add_row(['a','blah','boo'])
thing.add_row(['c','Im colored','right?'],['','red'])
print("Create 6th row that will have just a row attribute of class=my_class")
thing.add_row(['d','','Yay Multicolors'],['','','bg_brown,black'],{'html_row_attr': 'class="my_class"'})
thing.add_row(['b','hrmmmmmm','ahhh','stuffing','big empty'])
print("Rendering the HTML table")
thing.render()
del(rend)
del(thing)
